ukvm/ukvm_hv_openbsd_x86_64.c

```diff
@@ -0,0 +1,254 @@
+            .vrs_msrs[VCPU_REGS_STAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_LSTAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_CSTAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_SFMASK] = 0ULL,
+            .vrs_msrs[VCPU_REGS_KGSBASE] = 0ULL,
+            .vrs_crs[VCPU_REGS_XCR0] = XCR0_X87
+        }
+    };
+
+    ukvm_x86_setup_gdt(hv->mem);
+    ukvm_x86_setup_pagetables(hv->mem, hv->mem_size);
+
+    struct ukvm_boot_info *bi =
+        (struct ukvm_boot_info *)(hv->mem + X86_BOOT_INFO_BASE);
+    bi->mem_size = hv->mem_size;
+    bi->kernel_end = gpa_kend;
+    bi->cmdline = X86_CMDLINE_BASE; 
+    bi->cpu.tsc_freq = get_tsc_freq();
+
+    if (ioctl(hvb->vmd_fd, VMM_IOC_RESETCPU, &vrp) < 0)
+        err(1, "Cannot reset VCPU - exiting.");
+
+    *cmdline = (char *)(hv->mem + X86_CMDLINE_BASE);
+}
+
+void ukvm_hv_vcpu_loop(struct ukvm_hv *hv)
+{
+    struct ukvm_hvb         *hvb = hv->b;
+    struct vm_run_params    *vrp;
+    struct passwd *pw;
+    uid_t uid;
+    gid_t gid;
+
+    vrp = malloc(sizeof(struct vm_run_params));
+    if (vrp == NULL)
+        err(1, "malloc vrp");
+
+    vrp->vrp_exit = malloc(sizeof(union vm_exit));
+    if (vrp == NULL)
+        err(1, "malloc vrp_exit");
+
+    vrp->vrp_vm_id = hvb->vcp_id;
+    vrp->vrp_vcpu_id = hvb->vcpu_id;
+    vrp->vrp_continue = 0;
+
+    if ((pw = getpwnam(VMD_USER)) == NULL)
+        err(1, "can't get _vmd user");
+    uid = pw->pw_uid;
+    gid = pw->pw_gid;
+
+    if (chroot(pw->pw_dir) == -1)
+        err(1, "chroot: %s", pw->pw_dir);
+
+    if (chdir("/") == -1)
+        err(1, "chdir(\"/\")");
+
+    if (setgroups(1, &gid) ||
+        setresgid(gid, gid, gid) ||
+        setresuid(uid, uid, uid))
+        err(1, "unable to revoke privs");
+
+    /*
+     * pledge in the vm processes:
+     * stdio - for malloc and basic I/O including events.
+     * vmm - for the vmm ioctls and operations.
+     */
+    if (pledge("stdio vmm", NULL) == -1)
+        err(errno, "pledge");
+
+    for (;;) {
+        vrp->vrp_irq = 0xFFFF;
+
+        if (ioctl(hvb->vmd_fd, VMM_IOC_RUN, vrp) < 0) {
+            if(errno == EIO) // guest halted with interrupts disabled, ie cli, hlt
+                return;
+
+            err(1, "ukvm_hv_vcpu_loop: vm / vcpu run ioctl failed");
+        }
+
+        /* If the VM is terminating, exit normally */
+        if (vrp->vrp_exit_reason == VM_EXIT_TERMINATED) {
+            return;
+        }
+
+        union vm_exit *vei = vrp->vrp_exit;
+
+        if (vrp->vrp_exit_reason != VM_EXIT_NONE) {
+            switch (vrp->vrp_exit_reason) {
+                case VMX_EXIT_INT_WINDOW:
+                case VMX_EXIT_CPUID:
+                case VMX_EXIT_EXTINT:
+                case SVM_VMEXIT_INTR:
+                case VMX_EXIT_EPT_VIOLATION:
+                case SVM_VMEXIT_NPF:
+                case SVM_VMEXIT_MSR:
+                case SVM_VMEXIT_CPUID:
+                    // nothing to be done here, as per vmd
```

Comment:
Is there _really_ nothing to be done here? Do we get these exits in normal operation? If not, they should be unhandled exits (same as unknown exit).

Reply:
I was following what vmd does [vcpu_exit#L1471-L1479](https://github.com/openbsd/src/blob/7e7a8c998653110b57ed8bc9fdb7cb231e91d0a8/usr.sbin/vmd/vm.c#L1471-L1479)

Reply:
base on the comment (L1472-L1478)
```
We may be exiting to vmd to handle a pending interrupt but
at the same time the last exit type may have been one of
these. In this case, there's nothing extra to be done
here (and falling through to the default case below results
in more vmd log spam).
```
these are normal operation. Would you like me to chase it down further with mlarkin@openbsd?

Reply:
Given that we don't use interrupts for anything, based on that comment I'd be inclined to remove these (turn them into an unhandled exit). It's easy enough to try, e.g. see if test_ping_serve runs fine with the change.